cypress-io/cypress
# feat: show cloud connect options when hoving spec list fields

- Closes https://github.com/cypress-io/cypress/issues/22932

### User facing changelog

Hovering over the "Latest runs" or "Average duration" columns in the Specs list will show buttons to help connect a project to the Cypress dashboard.

### Additional details


#### Implementation details
1. Added additional Button variant to match the styles in the design called "LinkBold"
2. Rendered new Button variant in component test plus added other existing variants that were not represented so they would be caught in the Percy snapshot
3. Added a new component called `SpecListCloudButton` to encapsulate the logic for the different states of a project being connected to the dashboard and show an appropriate CTA to help the user continue the workflow.  This logic to calculate this cloud connection status for the project was repeated from other components such as the `SpecHeaderCloudDataTooltip` and the `SpecsListBanners`.  This logic can be extracted to use the work being performed in #23379 that will encapsulate this cloud status in one place.
4. Added a new component called 'SpecListHoverCell' to encapsulate the logic for hovering over a cell in the Specs list to show the new button.  This logic does have some location dependent details in it related to styling to make it work in the Specs list.  To use this in a different location in the app, those details would need to be refactored into props.
    * I was not able to use the TailwindCss `group-hover` because the Spec row is already using `group` to change the icon when hovering the row.  Used the `useElementHover` from VueUse instead.
6. Changed grid columns for Specs list columns to use a flexible first column for the spec name and path and fixed widths for the last three columns since the width of the data for those columns is known.  This was done in order to make sure there was enough space for the new buttons that appear in the column on hover.  The columns do reduce in width and the button text for the new buttons is shortened at smaller breakpoints.
7. Updated the placeholder in the "Latest runs" column when the project is not connected to show two dashes (i.e. "--") just like the "Average duration" column

_Question for any reviewers_

- In order to defeat the hover functionality in the SpecListHoverCell component when not needed, I am used a 'prop' that just returns out of the `watch` function if set to true.  This value is calculated by the parent component.  Would it have been better to just conditionally replace that component with a plain `div`?

### Steps to test


Testing this feature will be similar to testing the Spec list banners ( #23256 ) and for the tooltips shown for the "Latest runs" and "Average duration" columns.  For each state shown in the table, confirm that the appropriate button is shown and clicking it performs the appropriate action.

| Cloud state | Button shown when hovering | Button action |
| -- | -- | -- |
| not logged in | "Connect" with user icon | opens login modal |
| logged in and project not connected | "Connect project" with  link icon | opens connect project modal |
| logged in and  project connected but no user access | "Request access" with play icon | requests access for the current user |
| logged in and request access has been sent | "Request sent" with play icon and button is disabled | No action for button |
| Project is fully connected to the cloud | No button shown when hovering | | 

At smaller breakpoints when the column headers for "Latest runs" and "Average duration" are shortened, then the button text will also be shortened to one word:

| Long text | Shortened text |
| -- | -- |
| Connect | _not shortened_ |
| Connect project | Connect |
| Request access | Request |
| Request sent | Sent |

### How has the user experience changed?


#### Column widths for Specs list have been updated to fit content better

![SpecList_NewColumnWidths](https://user-images.githubusercontent.com/1702361/187093646-70d337b3-20dd-42a1-918a-46746a363a67.gif)

#### New buttons shown when hovering columns in Specs list

| When not logged in |
| ---- |
| ![SpecListHoverButton_NotLoggedIn](https://user-images.githubusercontent.com/1702361/187093474-14f169a9-9af7-4f0d-9fd0-f966346328a5.gif) |

| When not connected to a project  |
| ---- |
|  ![SpecListHoverButton_ConnectProject](https://user-images.githubusercontent.com/1702361/187093569-3a2ab550-72ca-415e-877a-db4ed1be1187.gif) |

| When needing to request access to a project |
| -- |
| ![SpecListHoverButton_RequestAccess](https://user-images.githubusercontent.com/1702361/187093592-f0d47ca2-e457-4f41-86d2-c2f204806dee.gif) |

| When request for access has been sent |
| -- |
| ![SpecListHoverButton_RequestSent](https://user-images.githubusercontent.com/1702361/187093610-ae937fb7-d8ec-4cc1-8771-60bd77f43837.gif) |



### PR Tasks


- [NA] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
